In [1]:
from IPython.core.display import HTML
with open ("../style.css", "r") as file:
    css = file.read()
HTML(css)

# Linear Regression with SciKit-Learn

We import the module `pandas`.  This module implements so called <em style="color:blue;">data frames</em> and is more convenient than the module `csv` when reading a <tt>csv</tt> file. 

In [2]:
import pandas as pd

The data we want to read is contained in the <tt>csv</tt> file `'cars.csv'`.  

In [3]:
cars = pd.read_csv('cars.csv')
cars

mpg  cyl  displacement     hp  weight   acc  year  \
0    18.0    8         307.0  130.0  3504.0  12.0    70   
1    15.0    8         350.0  165.0  3693.0  11.5    70   
2    18.0    8         318.0  150.0  3436.0  11.0    70   
3    16.0    8         304.0  150.0  3433.0  12.0    70   
4    17.0    8         302.0  140.0  3449.0  10.5    70   
..    ...  ...           ...    ...     ...   ...   ...   
387  27.0    4         140.0   86.0  2790.0  15.6    82   
388  44.0    4          97.0   52.0  2130.0  24.6    82   
389  32.0    4         135.0   84.0  2295.0  11.6    82   
390  28.0    4         120.0   79.0  2625.0  18.6    82   
391  31.0    4         119.0   82.0  2720.0  19.4    82   

                          name  
0    chevrolet chevelle malibu  
1            buick skylark 320  
2           plymouth satellite  
3                amc rebel sst  
4                  ford torino  
..                         ...  
387            ford mustang gl  
388                  vw pickup  
389              dodge rampage  
390                ford ranger  
391                 chevy s-10  

[392 rows x 8 columns]

We want to convert the columns containing `mpg` into one **NumPy** array, while the remaining numerical attributes should be collected into a 
<em style="color:blue">feature matrix</em>.  

In [4]:
import numpy as np

In [5]:
X = np.array(cars[['cyl', 'displacement', 'hp', 'weight', 'acc', 'year']])
Y = np.array(cars['mpg'])

Let us inspect the first five rows of the matrix `X`.

In [6]:
X[:5]

array([[   8. ,  307. ,  130. , 3504. ,   12. ,   70. ],
       [   8. ,  350. ,  165. , 3693. ,   11.5,   70. ],
       [   8. ,  318. ,  150. , 3436. ,   11. ,   70. ],
       [   8. ,  304. ,  150. , 3433. ,   12. ,   70. ],
       [   8. ,  302. ,  140. , 3449. ,   10.5,   70. ]])

Since *miles per gallon* is in a reciprocal relation to the *fuel consumption*, we convert `Y` to its inverse.

In [7]:
Y = 1 / Y

We import the `linear_model` from **SciKit-Learn**:

In [8]:
import sklearn.linear_model as lm

We create a *linear model* for linear regression.

In [9]:
M = lm.LinearRegression()

We train this model using the data that we have read.

In [10]:
%%time
M.fit(X, Y)

CPU times: user 7.96 ms, sys: 2.31 ms, total: 10.3 ms
Wall time: 4.73 ms


LinearRegression()

The model `M` represents a linear relationship between the dependent variable $1/\texttt{mpg}$ and the independent variables $\texttt{cyl}$, $\texttt{displacement}$, $\texttt{hp}$, $\texttt{weight}$, $\texttt{acc}$, and $\texttt{year}$ of the form
$$\displaystyle \frac{1}{\texttt{mpg}}
              = \vartheta_0 + \vartheta_1 \cdot \texttt{cyl} 
              + \vartheta_2 \cdot \texttt{displacement} 
              + \vartheta_3 \cdot \texttt{hp}
              + \vartheta_4 \cdot \texttt{weight}
              + \vartheta_5 \cdot \texttt{acc}
              + \vartheta_6 \cdot \texttt{year}  
$$
We proceed to extract the coefficients $\vartheta_i$ for $i\in\{1,\cdots,6\}$.

In [11]:
ϑ0 = M.intercept_
ϑ0

0.08952960948549937

In [12]:
ϑ1, ϑ2, ϑ3, ϑ4, ϑ5, ϑ6 = M.coef_
ϑ1, ϑ2, ϑ3, ϑ4, ϑ5, ϑ6

(0.0013920589695842662,
 -1.702830824431511e-05,
 0.0001137563629386943,
 1.109095941069281e-05,
 0.0003388008470551787,
 -0.0012659672279685468)

Let us check how much of the variance is explained by our model.  This is done using the method `score`.

In [13]:
R2 = M.score(X, Y)
R2

0.8838956283269112

The linear model explains $88\%$ of the variation of the fuel efficiency.  In order to derive a better model, we would need both the <em style="color:blue;">reference area</em> of the car and the <em style="color:blue;">drag coefficient</em>.